### PR TITLE
fix: fix infinite refresh occuring when app does not allow guest access

### PR DESCRIPTION
**Pull request by thaddmt**
#### Description of changes
- When user refreshes credentials and receives an error (often can occur if signing out and not having guest access) the credentials object will be set to some error value without an expiration date. This was causing set timeout to be called instantly as we were not passing a number
  - https://github.com/aws-amplify/amplify-ui/issues/3367
- Added a check to ensure that expiration is a number that exists 
- Still needs unit and integration tests

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added]
- [ ] Relevant documentation is changed or added (and PR referenced)

### CODE REVIEW

File: __tests__/AmplifyMapLibreRequest.test.ts
Comment (0618) [inline __tests__/AmplifyMapLibreRequest.test.ts:77]: Thanks for adding the test!

File: src/AmplifyMapLibreRequest.ts
Comment (ioanabrooks) [inline src/AmplifyMapLibreRequest.ts:91]: `refreshCredentialsWithRetry` is the handler for both `signOut` and `refreshToken` Hub events, and since these are async, we could still have a race condition scenario, of `refreshToken` being handled at same time as `signOut` but `signOut` resolves faster and an error is thrown on refreshToken retry. No infinite loop but we could still see an error, correct me if I am wrong. My question here is why can't we avoid getting credentials on signOut?

Comment (thaddmt) [reply src/AmplifyMapLibreRequest.ts:91]: @ioanabrooks We can avoid getting credentials on signOut certainly but the original reason for the change was that if a user was showing a map and went from signed in user to a guest user, they should no longer have access to call the map API. I've tested this in an example app recently and after signing out I could no longer make further requests to get map tiles (I could still see the parts of the map I loaded though).

Also I believe the refreshToken event is only fired when we attempt to get new identity token and access token. As far as I'm aware that is not the same scenario as a signout or signin event. Though there is a case where a user signsOut then instantly attempts a refreshToken. In that case we would throw an error on the attempted refreshToken as we will also try to refreshCredentials.

Comment (ioanabrooks) [reply src/AmplifyMapLibreRequest.ts:91]: Got it, so maybe we can mitigate this by handling `tokenRefresh_failure` hub event. what do you think?

Comment (thaddmt) [reply src/AmplifyMapLibreRequest.ts:91]: Yep we would just need to verify that, I can see I even left a comment that says that refreshToken is not currently fired - https://github.com/aws-amplify/maplibre-gl-js-amplify/pull/194/files#diff-3962f2f1d3b5c5b6449fb8fbed0fe5a2a00142d04fb3449107d130ea9f577aa9R88

Comment (ioanabrooks) [reply src/AmplifyMapLibreRequest.ts:91]: I'd still like to see the `tokenRefresh_failure` event handled, and if the token refresh event does not get dispatched as expected, we should include a comment referencing reference why that is for future maintainers, open a github issue if one does not exist because users can customize the refresh rate to lower or greater than an hour on Cognito side.

**Review by ioanabrooks — APPROVED** [review]
Approving with the understanding that a backlog ticket will be created to address comments.

**Review by 0618 — APPROVED** [review]

**Review by sreeramsama — APPROVED** [review]

**Pull request merged**
